drivers/bluetooth/hci/apollox_blue.c

```diff
@@ -0,0 +1,320 @@
+
+		if (ret) {
+			LOG_ERR("SPI read error %d", ret);
+			break;
+		}
+	} while (0);
+
+	return ret;
+}
+
+bool bt_apollo_vnd_rcv_ongoing(uint8_t *data, uint16_t len)
+{
+	/* The vendor specific handshake command/response is incompatible with
+	 * standard Bluetooth HCI format, need to handle the received packets
+	 * specifically.
+	 */
+	if (am_devices_cooper_get_initialize_state() != AM_DEVICES_COOPER_STATE_INITIALIZED) {
+		am_devices_cooper_handshake_recv(data, len);
+		return true;
+	} else {
+		return false;
+	}
+}
+
+int bt_hci_transport_setup(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	int ret;
+
+	/* Configure the XO32MHz and XO32kHz clocks.*/
+	clock_control_configure(clk32k_dev, NULL, NULL);
+	clock_control_configure(clk32m_dev, NULL, NULL);
+
+	/* Enable XO32kHz for Controller */
+	clock_control_on(clk32k_dev, (clock_control_subsys_t)CLOCK_CONTROL_AMBIQ_TYPE_LFXTAL);
+
+	/* Enable XO32MHz for Controller */
+	clock_control_on(clk32m_dev, (clock_control_subsys_t)CLOCK_CONTROL_AMBIQ_TYPE_HFXTAL_BLE);
+
+	/* Configure RST pin and hold BLE in Reset */
+	ret = gpio_pin_configure_dt(&rst_gpio, GPIO_OUTPUT_ACTIVE);
+	if (ret) {
+		return ret;
+	}
+
+	/* Configure IRQ pin and register the callback */
+	ret = gpio_pin_configure_dt(&irq_gpio, GPIO_INPUT);
+	if (ret) {
+		return ret;
+	}
+
+	gpio_init_callback(&irq_gpio_cb, bt_packet_irq_isr, BIT(irq_gpio.pin));
+	ret = gpio_add_callback(irq_gpio.port, &irq_gpio_cb);
+	if (ret) {
+		return ret;
+	}
+
+	/* Configure CLKREQ pin and register the callback */
+	ret = gpio_pin_configure_dt(&clkreq_gpio, GPIO_INPUT);
+	if (ret) {
+		return ret;
+	}
+
+	gpio_init_callback(&clkreq_gpio_cb, bt_clkreq_isr, BIT(clkreq_gpio.pin));
+	ret = gpio_add_callback(clkreq_gpio.port, &clkreq_gpio_cb);
+	if (ret) {
+		return ret;
+	}
+
+	/* Configure the interrupt edge for CLKREQ pin */
+	gpio_pin_interrupt_configure_dt(&clkreq_gpio, GPIO_INT_EDGE_RISING);
+
+	/* Take controller out of reset */
+	k_sleep(K_MSEC(10));
+	gpio_pin_set_dt(&rst_gpio, 0);
+
+	/* Give the controller some time to boot */
+	k_sleep(K_MSEC(500));
+
+	/* Configure the interrupt edge for IRQ pin */
+	gpio_pin_interrupt_configure_dt(&irq_gpio, GPIO_INT_EDGE_RISING);
+
+	return 0;
+}
+
+int bt_apollo_vnd_setup(spi_transmit_fun transmit)
+{
+	int ret;
+	struct net_buf *buf;
+	uint8_t *p;
+	am_devices_cooper_callback_t cb = {
+		.write = transmit,
+		.reset = bt_apollo_controller_reset,
+	};
+
+	/* Initialize the BLE controller */
+	ret = am_devices_cooper_init(&cb);
+	if (ret == AM_DEVICES_COOPER_STATUS_SUCCESS) {
+		am_devices_cooper_set_initialize_state(AM_DEVICES_COOPER_STATE_INITIALIZED);
+		LOG_INF("BT controller initialized");
+	} else {
+		am_devices_cooper_set_initialize_state(AM_DEVICES_COOPER_STATE_INITIALIZE_FAIL);
+		LOG_ERR("BT controller initialization fail");
+		return ret;
+	}
+
+	/* Set the NVDS parameters to BLE controller */
+	buf = bt_hci_cmd_create(HCI_VSC_UPDATE_NVDS_CFG_CMD_OPCODE,
```

Comment:
I understand that in very early stage, the device needed some configuration commands.
In this case, the driver calls higher level software api to self inject a command and handle reply.
Even though this is also used in the `spi.c` driver which inspired this code, I think this prevents compiling sample applications like `hci_uart` or similar that use hci raw access which does not provide support for api like `bt_hci_cmd_create`.
A possible solution for this problem is to use low level api provided by the driver itself (eg.  `bt_apollo_spi_send`), hence removing dependencies on `hci_core.c` file.

Please take this comment as point to discuss and not as request or recommendation for changes.

Reply:
In the case of building Zephyr as a pure controller + HCI transport with the help of `BT_HCI_RAW` I'd expect the host-side of the HCI transport (which could e.g. be Linux) to take over the responsibility of this kind of initialisation.

Reply:
That said, we could play with the idea of extracting out HCI command handling code from `hci_core.c` into a kind of library, which could optionally be included even in `BT_HCI_RAW` builds, however my initial feeling is that this is kind of violating what `BT_HCI_RAW` was created for, i.e. a simple and "dumb" pipe to the controller, so that another host can talk to it (and potentially also do early HCI init) instead of Zephyr.

Reply:
@jhedberg, I think it is a bit of grey area here, at least for me :). If we assume that Linux Host should now that a special command(s) needs to be sent to the device before starting operation, question is:

- How is Linux Host supposed to know about this?
- On the other hand, why Zephyr Host is not aware of this?
- What about connecting a chip running hci_uart (HCI raw) and a chip running Zephyr host ? [Obviously the Zephyr host will not know that some config commands are needed].

I was thinking that we can assume that some devices, normally designed for custom host implementation, requires some "setup" commands before starting standard operation. It is quite common to assume that this "setup" commands are sent using the standard hardware interface (SPI in our case) to avoid adding other physical interfaces between the two chips, so I would consider that during the setup phase, the device is not fully operational and so the regular hci command (even vendor specific) are not guaranteed to work.

If this agreed, then we can think of a setup phase (already present) where all these special configuration commands are sent, without relying on higher level software. After the setup phase, then regular HCI command (standard or vendor specific) needs to be supported.

In summary, we can simply forbid to use things like `bt_hci_cmd_send_sync` or similar during the setup phase or in general in all the driver.

Reply:
> Hi @msmttchr , thanks for your comment. Yes we always need samples for hci raw access for DTM or other purpose. It will have `bt_hci_cmd_create` compiling issue for that samples. I add `CONFIG_BT_HCI_RAW` condition here to add support for hci raw access. I would like to keep using `bt_hci_cmd_create` and `bt_hci_cmd_send_sync` for the vendor specific commands in non-raw case to make sure they are sent successfully since there is expected semaphore inside of these APIs to control.

@aaronyegx, I think adding `CONFIG_BT_HCI_RAW` condition is a good short-term approach to address the issue.

Reply:
@jhedberg , in some cases the "host" is just one serial tool or test machine which is requested to send the hci sequences and it has less knowledge about the controller running in the platform. It may be unavailable to do any initialization for the hci transport of Zephyr platform. The Bluetooth stack hci layer defines hci_core.c and hci_raw.c for general hci and hci raw access APIs, the hci driver here above the hci stack should be the best to decide whether the hci core or raw APIs to be executed in current framework.

Reply:
@jhedberg, thanks for the clarification, I think I understand a little bit more 😃 now.
I saw the implementation done by @aaronyegx which could be a good compromise at the end.